Un-merged PyGRB post-processing development

pannarale opened the pull request:
This PR refreshes work that had been carried out for the PyGRB post-processing and results presentation, but was never brought to master.

a-r-williamson commented on bin/pygrb/pycbc_pygrb_plot_injs_results:
```diff
@@ -0,0 +1,623 @@
+x_label = axis_labels_dict[x_qty]
+y_label = axis_labels_dict[y_qty]
+fig = plt.figure()
+xscale = "linear"
+if opts.x_log:
+    xscale = "log"
```
```suggestion
xscale = "log" if opts.x_log else "linear"
```

pannarale replied:
Hopefully I caught all of these.

a-r-williamson commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,868 @@
+from scipy import stats
+from pycbc.detector import Detector
+# TODO: imports to fix/remove
+try:
+    from glue import segments
+    from glue.ligolw import utils, lsctables, ligolw, table
+    from glue.segmentsUtils import fromsegwizard
+except ImportError:
+    pass
+try:
+    from pylal import MultiInspiralUtils
+except ImportError:
+    pass
```
`glue` and `pylal` need replaced

pannarale replied:
Here I switched all `glue` imports accordingly, but I am not sure I can do much about the `pylal` one.  We had encapsulated the `pylal` imports in the post-processing in `try...except` structures so the automatic tests would not be affected and with the plan to remove them as we developed things further.  This one up here is our last `pylal` import in the post-processing and there are no `MultiInspiralUtils` in `ligo` and its subpackages, so I would not know how to replace it.  Once we move to `hdf5`, this will become an obsolete dependency.

Btw, there are a couple more `pylal` imports in PyCBC.
1.  In `pycbc/results/legacy_grb.py`, but this file will be removed one the post-processing development is completed.  None of the new scripts or methods rely on it.
2. In `bin/pygrb/pycbc_make_offline_grb_workflow` when the post-processing is set up, and similarly to the above this part of the workflow will be soon superseded.
3. In `bin/pycbc_fit_sngl_trigs`.

@spxiwh, @ahnitz, can we leave the lines 42-45 @a-r-williamson quoted?  Ultimately they are already a narrowed down version of https://github.com/gwastro/pycbc/blob/master/pycbc/results/pygrb_plotting_utils.py#L38-L42 (and per this PR pygrb_plotting_utils.py will no longer have those lines).

spxiwh replied:
I didn't think pylal had ever been updated for python3, so is there ever a use case where this *can* do anything useful?

pannarale replied:
Right now that would only be running with Python2.7 to reproduce old results.  That's a way of ensuring the post-processing scripts do not break things in terms of analyses.

Should I move https://github.com/gwastro/pycbc-pylal/blob/master/pylal/MultiInspiralUtils.py#L66 into `pygrb_postprocessing_utils.py`?  Well, after readapting it and removing parts that are not needed.

I can't think of any other solution.  Moving ahead with the xml->hdf5 work with this PR would just make it too big in my opinion.

spxiwh replied:
But PyCBC now does not work *at all* on python 2.7. Older versions of PyCBC are needed if trying to reproduce old results.

pannarale replied:
I've commented on this in a summary message below (March 4, 2022) and we should continue the discussion from there.  In any event, all the code lines pointed out by @a-r-williamson are now updated.

a-r-williamson reviewed (dismissed):
@pannarale thanks for going through this and preparing a PR. There are a few things that jump out, especially imports of `glue` and `pylal`.

pannarale commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,691 @@
+from pycbc.results.pygrb_postprocessing_utils import pygrb_initialize_plot_parser
+from pycbc.results.pygrb_postprocessing_utils import pygrb_add_injmc_opts
+from pycbc.results.pygrb_postprocessing_utils import pygrb_add_bestnr_opts
+from pycbc.results.pygrb_postprocessing_utils import pygrb_add_fminjs_input_opts
```
Did this where there were more than 2 lines of `import pycbc.results.pygrb_postprocessing_utils...` and/or of `import pycbc.results.pygrb_plotting_utils...`

pannarale commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,691 @@
+Determine efficiency and exlucion distances of a PyGRB run.
+"""
+
+from __future__ import division
```
Removed these in the GRB stuff (but there are others lingering around in PyCBC).

pannarale commented on bin/pygrb/pycbc_pygrb_grb_info_table:
```diff
@@ -0,0 +1,120 @@
+description = 'Create GRB info table.'
+usage = __program__ + ' [--options]'
+
+parser = argparse.ArgumentParser(usage=usage, description=description,
```
@ahnitz, @spxiwh, before I follow up on this, can you confirm that it is indeed the convention we want?  I ask because I poked around and saw conflicting approaches within PyCBC.  I would rather make a single swipe :-)

pannarale replied:
`__doc__` is now used in all these post-processing scripts and `usage` is gone.

pannarale commented on bin/pygrb/pycbc_pygrb_plot_coh_ifosnr:
```diff
@@ -20,189 +20,249 @@
+if ifo is None:
     msg = "Please specify an interferometer"
     logging.error(msg)
     sys.exit()
```
I raised errors and I kept the `logging.error(...)`: let me know if you think that is redundant.

spxiwh commented on pycbc/results/pygrb_plotting_utils.py:
```diff
@@ -24,447 +24,26 @@
-import matplotlib
 if 'matplotlib.backends' not in sys.modules:  # nopep8
+    import matplotlib
     matplotlib.use('agg')
```
Please don't set this in any library code (as it can cause issues for people running Jupyter). Please put this in any executable that needs it instead.

spxiwh commented on bin/pygrb/pycbc_pygrb_pp_workflow:
```diff
@@ -24,165 +24,324 @@
+from pycbc.workflow import pegasus_workflow as wdax
+from Pegasus import DAX3 as dax
+from six.moves.urllib.request import pathname2url
+from six.moves.urllib.parse import urljoin
```
Please don't use `six` any more. Please just use the python3 versions.

pannarale replied:
I've commented on this in a summary message below (March 4, 2022)

spxiwh commented on bin/pygrb/pycbc_pygrb_pp_workflow:
```diff
@@ -24,165 +24,324 @@
+from pycbc.results import create_versioning_page, layout, pygrb_postprocessing_utils
 from pycbc.results.versioning import save_fig_with_metadata
+from pycbc.workflow import pegasus_workflow as wdax
+from Pegasus import DAX3 as dax
```
This is another "Old code needs updating before it can work on master" problem. We moved to pegasus5 as part of the python3 transition, but you are calling the old `DAX3` module. While this does still exist in pegasus, it will soon be removed. We are now using `Pegasus.api` as recommended. As `pegasus_workflow` has moved over to this, this code will be broken in many places. I'm reluctant to use the main pycbc repo. as a repository for broken code that is being promised to be fixed later, so it would be good to demonstrate that this branch can actually do a GRB PP workflow?

pannarale replied:
I've commented on this in a summary message below (March 4, 2022).

spxiwh commented on pycbc/workflow/plotting.py:
```diff
@@ -417,9 +419,11 @@ def make_results_web_page(workflow, results_dir, explicit_dependencies=None):
     node.add_opt('--template-file', template_path)
     workflow += node
     if explicit_dependencies is not None:
+        import Pegasus.DAX3 as dax
```
Here as well, we're not using pegasus4 any more.

pannarale replied:
Thanks, @spxiwh.  I looked at the changes to this file, and some of them, e.g., the one above, predate PR #3826, so they should not be there.  Perhaps there was an oversight in fixing conflicts when I rebased.  In any event, I undid these changes.

pannarale replied:
By the way, is 'depend**a**ncy' intentional in the various 'add_*_dependancy'?

spxiwh replied:
Whoops, that was a spelling mistake by me, but it is what the function is currently called https://pycbc.org/pycbc/latest/html/pycbc.workflow.html#pycbc.workflow.pegasus_workflow.Workflow.add_explicit_dependancy

spxiwh reviewed (changes requested):
Some requests from me as well! I do worry that a bunch of this doesn't work as is, and am a bit hesitant to commit "known to be broken" code.

titodalcanton commented on bin/pygrb/pycbc_grb_trig_cluster:
```diff
@@ -187,10 +185,10 @@ with h5py.File(args.trig_file, "r") as h5f:
-        print("copied input file to {}".format(outfile))
-    sys.exit(0)
+        print(err_msg)
+    raise RuntimeError(err_msg)
```
You can avoid printing a message if you are going to raise an exception with the same message right after.

titodalcanton commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,672 @@
+# =============================================================================
+# Function to calculate the error bars and fraction of recovered injections
+# (used for efficiency/distance plots)
+# =============================================================================
```
I would rather put this kind of descriptive comment into the function's docstring, so it ends up in the documentation.

titodalcanton commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,672 @@
+if opts.veto_directory and (opts.veto_category is None):
+    err_msg = "Must supply veto category if applying vetoes."
+    logging.error(err_msg)
+    raise RuntimeError(err_msg)
```
When you are validating command-line options, this is the preferred way:
```
if opts.veto_directory and (opts.veto_category is None):
    err_msg = "Must supply veto category if applying vetoes."
    parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,672 @@
+else:
+    err_msg = "Must provide both found and missed file if running injections."
+    logging.error(err_msg)
+    raise RuntimeError(err_msg)
```
```suggestion
    parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_efficiency:
```diff
@@ -0,0 +1,672 @@
+            logging.info("%s: sigmasq not set for at least one trigger.", ifo)
+        if sum(inj_sigma[ifo] != 0) == 0:
+            logging.info("%s: sigmasq not set for any trigger.", ifo)
+            if len(ifos) == 1:
+                msg = "This is a single ifo analysis. "
+                msg += "Setting sigmasq to unity for all triggers."
+                logging.info(msg)
+                inj_sigma[ifo][:] = 1.
```
Looks like this block can stay outside of the `for` loop?

pannarale replied:
But, we do not want to set `inj_sigma[ifo][:] = 1.` unless `sum(inj_sigma[ifo] != 0) == 0` (sigma-squared values are not provided for any trigger) and `len(ifos) == 1` (we are dealing with a single ifo).  By moving this block outside the `for`, we would set `inj_sigma[ifo][:] = 1.` every time we have a single ifo analysis.

titodalcanton replied:
Ok I see, I just misinterpreted the intent. Thanks.

titodalcanton commented on bin/pygrb/pycbc_pygrb_grb_info_table:
```diff
@@ -0,0 +1,115 @@
+
+opts = parser.parse_args()
+
+level = logging.INFO if opts.verbose else logging.WARNING
+logging.basicConfig(format="%(asctime)s: %(levelname)s: %(message)s",
+                    level=level)
```
```suggestion
pycbc.init_logging(opts.verbose, format="%(asctime)s: %(levelname)s: %(message)s")
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,216 @@
+import argparse
+import logging
+import h5py
+#from lal import GPSToUTC
```
```suggestion
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_page_tables:
```diff
@@ -0,0 +1,859 @@
+# Check options
+if opts.veto_directory and (opts.veto_category is None):
+    err_msg = "Must supply veto category if applying vetoes."
+    logging.error(err_msg)
+    raise RuntimeError(err_msg)
```
```suggestion
    parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_page_tables:
```diff
@@ -0,0 +1,859 @@
+
+    # Get trigs
+    on_trigs = ppu.load_xml_table(onsource_file, "multi_inspiral")
+    msg = "%d onsource triggers loaded." % (len(on_trigs))
+    logging.info(msg)
```
```suggestion
    logging.info("%d onsource triggers loaded.", len(on_trigs))
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_page_tables:
```diff
@@ -0,0 +1,859 @@
+    # Write to h5 file
+    if lont_h5_outfile:
+        logging.info("Writing loudest onsource trigger to h5 file.")
+        lont_h5_fp = h5py.File(lont_h5_outfile, 'w')
+        for i, key in enumerate(th):
+            lont_h5_fp.create_dataset(key, data=td[i])
+        lont_h5_fp.close()
```
```suggestion
        with h5py.File(lont_h5_outfile, 'w') as lont_h5_fp:
            for i, key in enumerate(th):
                lont_h5_fp.create_dataset(key, data=td[i])
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_page_tables:
```diff
@@ -0,0 +1,859 @@
+    # Write to h5 file
+    if qf_h5_outfile:
+        logging.info("Writing %d quiet-found injections to h5 file.", len(td))
+        qf_h5_fp = h5py.File(qf_h5_outfile, 'w')
+        for i, key in enumerate(th):
+            qf_h5_fp.create_dataset(key, data=td[i])
+        qf_h5_fp.close()
```
```suggestion
        with h5py.File(qf_h5_outfile, 'w') as qf_h5_fp:
            for i, key in enumerate(th):
                qf_h5_fp.create_dataset(key, data=td[i])
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_chisq_veto:
```diff
@@ -20,197 +20,332 @@
+    veto_string = ','.join([str(i) for i in range(2, opts.veto_category+1)])
     veto_files = glob.glob(opts.veto_directory +'/*CAT[%s]*.xml' %(veto_string))
-# If an ifo is given, the veto is intended as a single IFO quantity
+# Check that the y-axis variable was specified correctly
+if opts.y_variable in ['standard', 'bank', 'auto']:
+    veto_type = opts.y_variable
+else:
+    err_msg = 'Unrecognized option given with --y-variable'
+    logging.error(err_msg)
+    raise ValueError(err_msg)
+veto_type = opts.y_variable
```
This block is not necessary, just use the `choices` kwarg in `parser.add_argument()`.

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_chisq_veto:
```diff
@@ -20,197 +20,332 @@
+if use_sngl_ifo_snr:
+    if ifo is None:
+        err_msg = "--ifo must be given when using --use-sngl-ifo-snr"
+        logging.error(err_msg)
+        raise RuntimeError(err_msg)
+    else:
+        snr_type += '_sngl'
```
```suggestion
        parser.error(err_msg)
    snr_type += '_sngl'
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_chisq_veto:
```diff
@@ -20,197 +20,332 @@
+if use_sngl_ifo_veto:
+    if ifo is None:
+        err_msg = "--ifo must be given when using --use-sngl-ifo-snr-veto"
+        logging.error(err_msg)
+        raise RuntimeError(err_msg)
```
```suggestion
        parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_chisq_veto:
```diff
@@ -20,197 +20,332 @@
+        raise RuntimeError(err_msg)
+    elif veto_type != 'standard':
+        err_msg = "Single IFO values are available only for the standard chi-square"
+        logging.error(err_msg)
+        raise RuntimeError(err_msg)
```
```suggestion
        parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_coh_ifosnr:
```diff
@@ -20,189 +20,242 @@
-    msg = "Please specify an interferometer"
-    logging.error(msg)
-    sys.exit()
+if ifo is None:
+    err_msg = "Please specify an interferometer"
+    logging.error(err_msg)
+    raise RuntimeError(err_msg)
```
No need for this check, just use `required=True` in `add_argument()`.

titodalcanton replied:
Hmm after looking at the rest of the code, maybe what I suggested is not easy or convenient. Your choice, but at least
do the `parser.error(err_msg)` replacement as suggested above.

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_null_stats:
```diff
@@ -0,0 +1,237 @@
+    veto_string = ','.join([str(i) for i in range(2, opts.veto_category+1)])
+    veto_files = glob.glob(opts.veto_directory +'/*CAT[%s]*.xml' %(veto_string))
+
+# Check that the y-axis variable was specified correctly
+if opts.y_variable in ['coincident', 'nullstat', 'overwhitened']:
+    null_stat_type = opts.y_variable
+else:
+    err_msg = 'Unrecognized option given with --y-variable'
+    logging.error(err_msg)
+    raise ValueError(err_msg)
```
Same comment as above, use `choices` in `add_argument()`.

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_null_stats:
```diff
@@ -0,0 +1,237 @@
+
+# Overwhitened null stat (null SNR), null stat or coincident SNR vs
+# Coherent SNR plot
+if not opts.x_lims:
+    if zoom_in:
+        opts.x_lims = '6,30'
+if not opts.y_lims:
+    if zoom_in:
+        opts.y_lims = '0,30'
```
```suggestion
if not opts.x_lims and zoom_in:
    opts.x_lims = '6,30'
if not opts.y_lims and zoom_in:
    opts.y_lims = '0,30'
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_skygrid:
```diff
@@ -38,26 +39,46 @@ __date__ = pycbc.version.date
+    data['longitude'] = None
+    data['latitude'] = None
+
+    # Fill the dictionary in with required data
+    if input_file:
+        trigs = ppu.load_triggers(trig_file, vetoes)
+        data['longitude'] = numpy.degrees(trigs.get_column('ra'))
+        data['latitude'] = numpy.degrees(trigs.get_column('dec'))
+
+    return data
+
+
 # =============================================================================
 # Main script starts here
 # =============================================================================
 
 description = 'pycbc_pygrb_plot_skygrid will produce the sky grid plot.'
 usage = __program__ + ' [--options]'
-opts = pygrb_plot_opts_parser(usage=usage, description=description, version=__version__)
+parser = ppu.pygrb_initialize_plot_parser(usage=usage, description=description, version=__version__)
+opts = parser.parse_args()
 
-if opts.verbose:
-    level = logging.INFO
-else:
-    level = logging.WARNING
+level = logging.INFO if opts.verbose else logging.WARNING
 logging.basicConfig(format="%(asctime)s:%(levelname)s : %(message)s",
                     level=level)
```
```suggestion
init_logging(opts.verbose, format="%(asctime)s:%(levelname)s : %(message)s")
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_snr_timeseries:
```diff
@@ -20,131 +20,194 @@
+    veto_string = ','.join([str(i) for i in range(2, opts.veto_category+1)])
     veto_files = glob.glob(opts.veto_directory +'/*CAT[%s]*.xml' %(veto_string))
-snr_type = opts.variable
+if opts.y_variable in ['coherent', 'single', 'reweighted', 'null']:
+    snr_type = opts.y_variable
+else:
+    err_msg = 'Unrecognized option given with --y-variable'
+    logging.error(err_msg)
+    raise ValueError(err_msg)
```
Same comment as before, use `choices`.

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_stats_distribution:
```diff
@@ -0,0 +1,238 @@
+# Check options
+if opts.veto_directory and (opts.veto_category is None):
+    err_msg = "Must supply veto category if applying vetoes."
+    logging.error(err_msg)
+    raise RuntimeError(err_msg)
```
```suggestion
    parser.error(err_msg)
```

titodalcanton commented on bin/pygrb/pycbc_pygrb_plot_stats_distribution:
```diff
@@ -0,0 +1,238 @@
+if not opts.newsnr_threshold:
+    opts.newsnr_threshold = opts.snr_threshold
+
+if opts.x_variable not in ["bestnr", "snr", "snruncut"]:
+    err_msg = "--x-variable must be one of: bestnr, snr, snruncut."
+    logging.error(err_msg)
+    raise ValueError(err_msg)
```
Same comment as before, use `choices`.

titodalcanton commented on bin/pygrb/pycbc_pygrb_pp_workflow:
```diff
@@ -24,165 +24,324 @@
-    #inj_file.ifo_list=wflow.ifos
+        node.new_output_file_opt(workflow.analysis_time, '.png',
+                                 '--output-file', tags=extra_tags)
+        if exec_name in ['pygrb_plot_coh_ifosnr', 'pygrb_plot_null_stats']:
+            if 'zoomin' in tags:
+                node.add_opt('--zoom-in')
```
```suggestion
        if exec_name in ['pygrb_plot_coh_ifosnr', 'pygrb_plot_null_stats'] and 'zoomin' in tags:
            node.add_opt('--zoom-in')
```

pannarale commented:
@a-r-williamson, @spxiwh, @titodalcanton, thanks for looking through this material and all the suggestions/requests you provided.  Let me summarise things here, and raise an issue myself, so we do not have to scroll through this page to keep iterating on this.

1. I have to clean up some things raised by codeclimate.
2. There is one request that I have not worked on yet, and that is @spxiwh's one about `matplotlib`
3. @spxiwh, I updated the workflow generator (`pycbc_pygrb_pp_workflow`) so that it successfully generates a workflow under Pegasus5. There is a "but" coming up later.
4. I downsized the last `pylal` function we needed and moved it over to `pycbc/results/pygrb_postprocessing_utils.py`. But...

`ligo.lw` does not support `MultiInspiralTable`!  So right now, all automatic tests are successful (minus code climate), I can generate a workflow, but there is a fundamental roadblock with `ligo.lw`.  Nothing is broken but nothing can work, in other words.

Ultimately we will not use `MultiInspiralTable`s but hdf5 files, however it will be useful to be able to post-process old results for a number of reasons (reviews, training, etc.).

Suggestions?

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+#parser.add_argument('--instrument', help="Name of interferometer e.g. H1")
+parser.add_argument('--wiki-file',
+                    help="Name of file to save wiki-formatted table in")
+parser.add_argument('--output-map')
```
If you give the `include_subdax_opts` option to `add_workflow_settings_cli` then you don't need this.

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+parser.add_argument('--wiki-file',
+                    help="Name of file to save wiki-formatted table in")
+parser.add_argument('--output-map')
+parser.add_argument('--transformation-catalog')
```
There is no longer a separate transformation catalog file, so this option should be removed.

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+                    help="Name of file to save wiki-formatted table in")
+parser.add_argument('--output-map')
+parser.add_argument('--transformation-catalog')
+parser.add_argument('--output-file')
```
I would also recommend to remove this, instead use the `--dax-file` option to set the output file name.

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+
+init_logging(args.verbose, format="%(asctime)s: %(levelname)s: %(message)s")
+
+workflow = wf.Workflow(args, args.workflow_name)
```
Then remove `args.workflow_name` here. (The `__init__` method of `Workflow` will default to looking for this argument anyway).

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+
+    layouts += list(layout.grouper(files, 2))
+
+workflow.save(filename=args.output_file, output_map_path=args.output_map)
```
And finally here just have `workflow.save()` 

The example for this is https://github.com/gwastro/pycbc/blob/master/bin/minifollowups/pycbc_foreground_minifollowup

spxiwh commented on bin/pygrb/pycbc_pygrb_pp_workflow:
```diff
@@ -16,185 +16,322 @@
+    node.add_opt('--wiki-file', wikifile)
+    if tags:
+        node.add_list_opt('--tags', tags)
+    node.new_output_file_opt(workflow.analysis_time, '.dax', '--output-file')
```
This would then also update to `--dax-file`. In this case the example is https://github.com/gwastro/pycbc/blob/master/pycbc/workflow/minifollowups.py#L100

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+num_events = min(num_events, len(fp['BestNR'][:]))
+
+# Determine ifos used in the analysis
+trig_file = workflow.cp.get('workflow', 'trig-file')
```
This is probably a FIXME for the future, but it's not best practice to send files through like this. It assumes a shared file system (which is normally true in practice, but doesn't have to be). Better to have this as an explicit input file. The workflow generator itself always runs on a head node, so can assume this safely, but this code will run on a compute node.

spxiwh commented on bin/pygrb/pycbc_pygrb_minifollowups:
```diff
@@ -0,0 +1,205 @@
+    node.add_opt('--trig-file', trig_file_path)
+    # Veto directory and segments directory
+    node.add_opt('--veto-directory', workflow.cp.get('workflow', 'veto-directory'))
+    node.add_opt('--segment-dir', workflow.cp.get('workflow', 'segment-dir'))
```
This is probably a FIXME for the future, but it's not best practice to send files through like this. It assumes a shared file system (which is normally true in practice, but doesn't have to be). Better to have this as explicit input files. The workflow generator itself always runs on a head node, so can assume this safely, but this code will run on a compute node .... As will work in most instances as is, this is probably low down the priority tree, but adding a FIXME into the code itself here might be a good idea.

spxiwh commented on bin/pygrb/pycbc_pygrb_pp_workflow:
```diff
@@ -16,185 +16,322 @@
+                node.add_opt('--'+opt, workflow.cp.get('workflow', opt))
+    # Injection file
+    if inj_set is not None and exec_name != 'pygrb_plot_injs_results':
+        node.add_opt('--inj-file', workflow.cp.get('injections-'+inj_set, 'inj-file'))
```
Another FIXME for the future: There's a lot of cases of input files being declared with `add_opt`. In this case pegasus does not know that this is a file, and cannot do any sort of transfer. On a shared filesystem this should always work, but if (for example) you have a large trigger file, and a job might parse through all of it multiple times, it is more efficient to have it copied to a local disk first. Pegasus can handle this for you, but it needs to know these are input files.

spxiwh reviewed (commented):
A few pointers from me on the workflow stuff. It's mostly suggestions for the future (some of these are nontrivial), but there are one or two things worth fixing now, which are all minor.

pannarale commented:
> A few pointers from me on the workflow stuff. It's mostly suggestions for the future (some of these are nontrivial), but there are one or two things worth fixing now, which are all minor.

Thanks, @spxiwh!  I followed up on all your comments and pushed the changes.

The one thing I did not update is the fact that some of these scripts take paths to the veto and segments directories as options: those are still handled as `add_opt` in the workflow generators, despite the fact that (some of) the files within those directories will be read and should be copied.  I left a FIXME in `pygrb_pp_workflow`.

pannarale commented:
> @a-r-williamson, @spxiwh, @titodalcanton, thanks for looking through this material and all the suggestions/requests you provided. Let me summarise things here, and raise an issue myself, so we do not have to scroll through this page to keep iterating on this.
> 
> 1. I have to clean up some things raised by codeclimate.
> 2. There is one request that I have not worked on yet, and that is @spxiwh's one about `matplotlib`
> 3. @spxiwh, I updated the workflow generator (`pycbc_pygrb_pp_workflow`) so that it successfully generates a workflow under Pegasus5. There is a "but" coming up later.
> 4. I downsized the last `pylal` function we needed and moved it over to `pycbc/results/pygrb_postprocessing_utils.py`. But...
> 
> `ligo.lw` does not support `MultiInspiralTable`! So right now, all automatic tests are successful (minus code climate), I can generate a workflow, but there is a fundamental roadblock with `ligo.lw`. Nothing is broken but nothing can work, in other words.
> 
> Ultimately we will not use `MultiInspiralTable`s but hdf5 files, however it will be useful to be able to post-process old results for a number of reasons (reviews, training, etc.).
> 
> Suggestions?

Going back the topic of `MultiInspiralTable`s, it turns out that `glue.ligolw` still supports these tables a the moment, so the easiest thing to do is to still use `glue.ligolw` whenever a `MultiInspiralTable` is involved.  This is the solution in place right now.  Removing `glue.ligolw` breaks the post-processing.

Again, there will be a move to hdf5 that will make this obsolete, so cooking up more involved ways of dealing with `MultiInspiralTable`s seems somewhat a waste of effort/time.

titodalcanton commented on bin/pygrb/pycbc_grb_trig_cluster:
```diff
@@ -187,10 +185,10 @@ with h5py.File(args.trig_file, "r") as h5f:
+    err_msg = "trigger file is empty\ncopied input file to {}".format(outfile)
     if args.verbose:
-        print("trigger file is empty")
-        print("copied input file to {}".format(outfile))
-    sys.exit(0)
+        print(err_msg)
+    raise RuntimeError(err_msg)
```
Looking at this again now, I realized that this is not an error condition, is it? In that case you should not raise an exception, otherwise HTCondor or whatever will consider the job as failed. I am not sure how frowned upon `sys.exit(0)` is, but it does look like a practical shortcut in this case… Maybe add a FIXME to be improved later?

titodalcanton replied:
And please do replace the if+`print()` with just a `logging.info()`.

pannarale replied:
True, it is not an error.   Avoiding the `sys.exit(0)` could be achieved by placing all the lines of code that follow in an `else`.  That is kind of ugly, though.

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+        segs = fromsegwizard(open(os.path.join(seg_dir, file_name), 'r'))
+        if len(segs) > 1:
+            err_msg = 'More than one segment, an error has occured.'
+            logging.error(err_msg)
```
```suggestion
```

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+            # If we got this far and the bestNR is non-zero, verify that chisq
+            # was actually calculated for the trigger, otherwise print
+            # debugging info
+            if bestnr[i_trig] != 0 and trig.chisq == 0:
+                err_msg = "Chisq not calculated for trigger with end time "
+                err_msg += f"{trig.get_end()} and SNR {trig.snr}."
+                logging.error(err_msg)
+                raise RuntimeError(err_msg)
```
Is the exception here intended? The comment claims this should just print debugging info, so maybe remove the exception and turn `logging.error()` into `logging.info()`?

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+                    continue
+                err_msg = "Triggers found in input files not in the list of "
+                err_msg += "analysed segments. This should not happen."
+                logging.error(err_msg)
```
```suggestion
```

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+    if not (numpy.all(ids[1:] == numpy.array(ids[:-1])+1) and ids[0] == 0):
+        err_msg = "time_slide_ids list should start at zero and increase by "
+        err_msg += "one for every element"
+        logging.error(err_msg)
```
```suggestion
```

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+        err_msg = "The zero-lag slide should have time_slide_id == 0 "
+        err_msg += "but the first element of time_slide_list is "
+        err_msg += f"{time_slide_list[0]}."
+        logging.error(err_msg)
```
```suggestion
```

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+                else:
+                    err_msg = 'zero_lag_slide_id was already assigned: there'
+                    err_msg += 'seems to be more than one zero-lag slide!'
+                    logging.error(err_msg)
```
```suggestion
```

titodalcanton commented on pycbc/results/pygrb_postprocessing_utils.py:
```diff
@@ -0,0 +1,905 @@
+    if zero_lag_slide_id is None:
+        err_msg = 'Unable to assign zero_lag_slide_id: '
+        err_msg += 'there seems to be no zero-lag slide!'
+        logging.error(err_msg)
```
```suggestion
```

spxiwh commented:
Thanks @pannarale, I think my comments have all been resolved now. I'll leave it to Tito to give final approval when ready.

titodalcanton reviewed (approved):
Thanks @pannarale, I have no further comments.

titodalcanton commented:
@a-r-williamson @spxiwh I see that your requested changes are still pending… Maybe you explicitly need to approve the PR to unblock that?

spxiwh commented:
@titodalcanton I can unblock the hold on this due to @a-r-williamson 's request, but can you first assure me that these changes have been made in response to this request?

titodalcanton commented:
@spxiwh yes, modulo the removal of `glue` imports which are still needed for reasons we discussed (and will be removed in short order as we switch to HDF).

spxiwh commented:
Okay, I removed the block, and looks like auto-merge did it's thing. If there were any pending concerns, these can be re-raised as issues for @pannarale to consider, but do want to help push this forward.
